### PR TITLE
Fixes #6 -- Create Data Ingestion Pipeline

### LINKED ISSUE #6: Create Data Ingestion Pipeline
The "data ingestion pipeline" should be built to perform these tasks:

1. Ingest a JSON file, containing parameters (according to the schema defined in https://github.com/sustainableaviation/EcoPyLot/issues/7)
2. Output a 4-dimensional [`xarray`](https://docs.xarray.dev/en/stable/), containing the dimensions (`pax`, `propulsion`, `year`, `parameters`).
3. Output a tuple of data structures, containing the _unique_ values of the 4 dimensions (`pax`, ...) 

In addition, the following utility functionality should be built:

1. Check for input data validity (valid JSON? corresponding to defined schema?)
2. logging outputs in every function, including statistics (eg. `file loaded, read 778 unique parameters`)

> [!NOTE]
> It is useful to think of the 4d `xarray` as a collection of 3d `xarray`s:
> 
> xarray

### PR DESCRIPTION
**Pull request by iamsiddhantsahu**
Implements a class `JSONToDataFrameConverter` with the method `create_dataframe()` that reads a JSON file in the specified format which is detailed in the documentation and outputs a Pandas DataFrame object.

⚠️ To Note:

1. ~Does not support reading XArray objects~ as it was initially planned in #6 because it was investigated that XArray consume a lot more memory that Pandas DataFrame.
2. It also does not support for output with `Stochastic` and `Static` variants of the DataFrame.

### CODE REVIEW

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import json
+import pandas as pd
+    
+class JSONToDataFrameConverter:
+    """
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:5]: Formatted according to the [`numpy` docstrings standard](https://numpydoc.readthedocs.io/en/latest/format.html) - great!

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import json
+import pandas as pd
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:2]: ...best to double-check if we are using any pandas APIs that could be deprecated in the future. If so, we should pin the version.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    
+class JSONToDataFrameConverter:
+    """
+    A class to convert JSON data into a Pandas DataFrame.
+
+    This class takes JSON input, either as a raw string or a path to a JSON file,
+    and converts it into a Pandas DataFrame. The JSON structure is expected to 
+    contain nested data, with one of the fields as a list that needs to be expanded
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:10]: > with one of the fields as a list that needs to be expanded

...what does that mean? Must be clear to users from the docstrings alone.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    into multiple rows in the DataFrame. Each key in the JSON object is considered a 
+    unique identifier (UID) for the rows.
+
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:13]: ...an example JSON and resulting dataframe should be added to the docstring here. It does not have to be the full JSON that you list in https://github.com/sustainableaviation/EcoPyLot/pull/13/files, but should show the logic of the function.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            A dictionary representation of the JSON data.
+
+        Raises
+        ------
+        FileNotFoundError
+            If the json_input is a file path and the file does not exist.
+        json.JSONDecodeError
+            If the json_input is not a valid JSON string.
+        """
+        if json_input.endswith('.json'):
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:68]: Would `json.loads(json_input)` not accept _both_ a json-like string and a file path anyway?

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        Loads JSON data from a file or a string.
+
+    create_dataframe():
+        Processes the JSON data and converts it into a Pandas DataFrame, expanding
+        lists into multiple rows and including other attributes in each row.
+
+    Parameters
+    ----------
+    json_input : str
+        A JSON string or a path to a JSON file containing the data to be converted.
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:26]: ...you could specify here that it accepts those inputs that the `json.load(file)` accepts. Link to the API docs of that function - then we are clear and unambigous.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+import json
+import pandas as pd
+    
+class JSONToDataFrameConverter:
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:4]: Is there any particular reason you are implementing this rather simple two-function logic as a class? Will we need to extend this somehow? What other class could possibly inherit from it?

Comment (michaelweinold) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: ...I am personally a big fan of using _native_ data types, and not extend them unnecessarily.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    Parameters
+    ----------
+    json_input : str
+        A JSON string or a path to a JSON file containing the data to be converted.
+
+    Attributes
+    ----------
+    json_data : dict
+        The JSON data loaded into a Python dictionary.
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:31]: > The JSON data loaded into a Python dictionary.

what is loaded into a Python dictionary here?

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                data = json.load(file)
+        else:
+            data = json.loads(json_input)
+        return data
+
+    def create_dataframe(self):
+        """
+        Creates and returns a Pandas DataFrame from the loaded JSON data.
+
+        This method processes the JSON data, expanding lists into multiple rows
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:79]: > expanding lists into multiple rows
         and incorporating other attributes in each row.

not clear enough.

File: EcoPyLot/JSONToDataFrameConverter.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        properties duplicated across these rows.
+
+        Returns
+        -------
+        pandas.DataFrame
+            A DataFrame containing the processed data from the JSON input.
+        """
+        frames = []
+        for uid, values in self.json_data.items():
+            for size in values['sizes']:
```
Comment (michaelweinold) [inline EcoPyLot/JSONToDataFrameConverter.py:91]: You do realize that other parameters could also be implemented as lists? It is only in our example in https://github.com/sustainableaviation/EcoPyLot/pull/13/files that `sizes` is the only parameter that contains a list.

File: EcoPyLot/dataloader.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
+import json
+import pandas as pd
+    
+class JSONToDataFrameConverter:
```
Comment (michaelweinold) [inline EcoPyLot/dataloader.py:4]: Now I'm confused - how is this connected to `def create_dataframe(self):` in `EcoPyLot/JSONToDataFrameConverter.py`?

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 7 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Yes, this is a valid point. All the `JSONToDataFrameConverter` class is doing is reading a JSON file or JSON string and generating a Pandas DataFrame. All the functionality could indeed be moved into a `JSONToDataFrameConverter()` function instead.

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 5 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:68]: This `if-else` condition is used to handle two different types of input that the function might receive: a JSON file path or a JSON string.

1. **If the input is a JSON file path:** The `endswith('.json')` condition checks if the input string ends with '.json', which is a common file extension for JSON files. If this condition is true, the code opens the file using `open(json_input)` and loads the JSON data from the file using `json.load(file)`.

2. **If the input is a JSON string:** If the input does not end with '.json', the function assumes that the input is a JSON string. It then parses the JSON string using `json.loads(json_input)`.

In summary, this `if-else` condition allows the function to handle both JSON files and JSON strings, making it more flexible.

Comment (michaelweinold) [reply EcoPyLot/JSONToDataFrameConverter.py:68]: Ah, the distinction is `load` vs `loadS`.
If I need your code summarized by ChatGPT, I'll do it myself, thank you very much.

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 7 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Should we move the functionality to just a function instead of a class?

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 2 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:2]: Yes, but this should be taken care by the `requirements.txt` file or the `environment.yml` file. Right?

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 7 above)
Comment (michaelweinold) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Good question - in this case, I had hoped that you would be able to provide some guidance. If there are good reasons to retain the current `class` structure, I have not found them. Feel free to elaborate - otherwise, let's go with a well-documented function instead.

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 2 above)
Comment (michaelweinold) [reply EcoPyLot/JSONToDataFrameConverter.py:2]: ...sure, as soon as we have one. The actual _thinking-about-pinning-the-version_ will need to happen in the code though. To me it seems we are using very stable, core parts of the API. I just had a few bad experiences with `numpy` recently (where I got depreciation warnings).

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 7 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Yes, I would say so too -- the operations performed by the `JSONToDataFrameConverter` class are not "complex" enough to make into a class. And, the class only has **one** method `create_dataframe()`. So it is better to have all these functionality done just by a function. I am converting this to a function instead.

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 5 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:68]: Reading the official documentation: 
1. [json.load()](https://docs.python.org/3/library/json.html#json.load) -- accepts only a file pointer to a file containing json data either in text format or binary format. 
2. Whereas, [json.loads()](https://docs.python.org/3/library/json.html#json.loads) -- accepts [str](https://docs.python.org/3/library/stdtypes.html#str), [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) or [bytearray](https://docs.python.org/3/library/stdtypes.html#bytearray)

So reading from the official documentation it seems like `json.loads()` cannot accept a file pointer i.e. a path to the file in the file system. But, I can give it a try if it is possible to read JSON files with the `loads()` method.

File: EcoPyLot/JSONToDataFrameConverter.py (hunk 7 above)
Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Do we plan to use any other data formats than JSON format in the future?

Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: I wanted to have the class called `EcoPyLot` that implements the Excel model described in the https://www.nature.com/articles/s41467-023-39749-y paper. Therefore, I wanted to keep the dataloader logic i.e. `JSONToDataFrameConverter` decoupled from the `EcoPyLot`. What are your views? @michaelweinold @romainsacchi

Comment (michaelweinold) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: It seems to me that a _class_ EcoPylot does not make sense from the perspective of the package. After all, we would ultimately (very simplified)

```python
import ecopylot as ep

mydf = ep.load_data(data.json)
myresults = ep.calculate_emissions(mydf)
```

In any case, this does not touch the question of whether the data ingestion logic should be implemented in a class. I don't think it should be. We can discuss this in more detail tomorrow - until then, you have plenty of comments to work on.

Comment (iamsiddhantsahu) [reply EcoPyLot/JSONToDataFrameConverter.py:4]: Okay thanks for sharing your thoughts.

This is how I see the package being used.
```python
import ecopylot as ep

ep_model_static = ep.EcoPyLot(type=static)
ep_model_stochastic = ep.EcoPyLot(type=stochastic, sample_size=2)

df_static = ep_model_static.load_data(data.json)
co2_emmissions_static = ep_model_static.calculate_emissions(df_static)

df_stochastic = ep_model_static.load_data(data.json)
co2_emmissions_stochastic = ep_model_static.calculate_emissions(df_stochastic)
```
What are your views on having this API? @michaelweinold @romainsacchi 

In terms of the project structure this is how I see the `ecopylot` package.
- Name of the Python package is `ecopylot` i.e. directory name is also the same `ecopylot` with `__init__.py` inside it.
- Inside this directory there is only one Python module exposing the `class EcoPylot`. This class implements the `load_data()` method responsible for loading JSON text file and converting them in to Pandas DataFrame. Similarly, this also implements the `calculate_emissions()` method.

🔔 This mean in the whole package there would be only one class called `EcoPylot` that implements all the functionality in this methods -- for simplicity.
